### ISSUE TITLE
RandomGrayScale prob is not for gray scale

### ISSUE THREAD

**Issue by guanfuchen** [jramapuram/datasets#4]
I use the [code](https://github.com/jramapuram/datasets/blob/7c5d0d9a1d78ddde6204571d1c39109f6da5e9d1/dali_imagefolder.py#L33), but the prob 0.2 is not for meaning transform gray scale in 0.2 prob.

![image](https://user-images.githubusercontent.com/22321977/103059673-bcf38980-45e0-11eb-8fd8-42d3d140c2df.png)

**Comment by youweiliang** [comment]
@guanfuchen Hi, glad to see you again. Have you figured out what is the fastest way to load ImageNet data for self-supervied augmentation as in BYOL/SimCLR/MoCo? I find some methods like using DALI, using [memcache](https://github.com/open-mmlab/OpenSelfSup/issues/15), torchvision Tensor augmentation, and others. But not sure which is better.

**Comment by youweiliang** [comment]
By the way, I think [nvidia.dali.ops.CoordTransform](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/supported_ops.html#nvidia.dali.ops.CoordTransform) can do the gray scale.

**Comment by jramapuram** [comment]
Hi @guanfuchen: I take it you mean the Dali ops (link to code) and not the torchvision operations (which you have a screenshot of)?

In this case this it should be correct. The coin flip is an RNG and it is used to parameterize the saturation of a hsv transform. See:

 - https://github.com/klecki/DALI/blob/c63f767e5d4a120c304a1a62393a30e71afb862b/docs/examples/image_processing/hsv_example.ipynb
 - https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/image_processing/hsv_example.html

**Comment by youweiliang** [comment]
I agree that the implementation of gray scale in the repository is correct. Now [GaussianBlur](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/supported_ops.html#nvidia.dali.ops.GaussianBlur) is also available in Dali. So I think Dali can be used to implement all the augmentations in BYOL.

**Comment by guanfuchen** [comment]
@youweiliang Hi, I have check the DALI dataloder in BYOL using other network structure, it is right now. I am waiting for tmorrow's result. DALI will get excellent speedup. Try it!!!

![image](https://user-images.githubusercontent.com/22321977/103280248-e3635b80-4a0a-11eb-99a5-057608872d6f.png)

**Comment by guanfuchen** [comment]
@jramapuram I try using prob 1.0 for whole gray scale image process. But the code performs different.

**Comment by youweiliang** [comment]
@guanfuchen Great! Thanks. I will have a try. DALI does seem better because it is written in C++, which I guess would speed up things a lot.

**Comment by guanfuchen** [comment]
@youweiliang you should be attentation with **the drop last**. Construct a small dataset(16 images) to check the dataloader performs like pytorch dataloader. I will be glad for some modification for your dataset repo, you do a great work for 
compatible with pytorch dataloader. I just use some code snippets for my dataloader. It is worth for speeding because I have a lot of experiments for trying

**Comment by youweiliang** [comment]
@guanfuchen Thanks for your great suggestions! I guess it does not matter much whether to drop last or not, right? By the way, why do you use `Solarize` in your augmentation, which is not used in BYOL? Does it help improve the performance?

**Comment by youweiliang** [comment]
Sorry, I just found BYOL uses `Solarize` in it augmentation.

**Comment by guanfuchen** [comment]
@youweiliang https://github.com/yaox12/BYOL-PyTorch/blob/master/data/byol_transform.py#L47 BYOL use the dataset operation, and you also add this. It will drop some accuracy for some transfer learning task.

**Comment by youweiliang** [comment]
Got it. Thanks!

**Comment by jramapuram** [comment]
@guanfuchen : not sure why this would be the case for RandomGrayScale --> it should be directly from the DALI ipynb. If not I believe this should be an issue opened against DALI :) 
@youweiliang: you can find the official transforms used [here](https://github.com/deepmind/deepmind-research/blob/master/byol/utils/augmentations.py#L27-L66). Note that different views use different transforms.

Let me know if there are any other issues directly related to this code :)

**Comment by zimenglan-sysu-512** [comment]
hi @jramapuram 
i use [RandomGrayScale(object)](https://github.com/jramapuram/datasets/blob/270e189393fedecb4e4df68ea59600e3005b67ec/dali_imagefolder.py#L33) to visualize the covert gray image by setting `prob=1.0`, but it does not look like gray.
![image](https://user-images.githubusercontent.com/13795888/203500552-cb1a3ffa-7ea0-45d6-86ac-c53bbb7a5862.png)

**Comment by zimenglan-sysu-512** [comment]
> hi @jramapuram i use [RandomGrayScale(object)](https://github.com/jramapuram/datasets/blob/270e189393fedecb4e4df68ea59600e3005b67ec/dali_imagefolder.py#L33) to visualize the covert gray image by setting `prob=1.0`, but it does not look like gray. ![image](https://user-images.githubusercontent.com/13795888/203500552-cb1a3ffa-7ea0-45d6-86ac-c53bbb7a5862.png)

it works when set prob=0 instead of 1.0, please see https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/image_processing/hsv_example.html

**Comment by jramapuram** [comment]
Good find @zimenglan-sysu-512 ! You need to change CoinFlip in RandomGrayscale to `1.0-prob` as per the doc you mentioned. Thanks for the report :)
